### ISSUE TITLE
PyTorch simulation example with GPUs

### ISSUE THREAD

**Issue by Mirian-Hipolito** [adap/flower#1415]
### What is your question?

Hello, I'm able to run the PyTorch simulation example (https://github.com/adap/flower/tree/main/examples/simulation_pytorch)  included in the repo using CPU, however I couldn't find an example of how to change the setup to use GPUs instead. Can you please provide some orientation of how to run this script using 4 GPUs? Is this possible? or the simulation is designed to use CPUs only?

**Comment by pedropgusmao** [comment]
Hi there, 

In reality, the example should be running on your first Nvidia GPU as seen here: https://github.com/adap/flower/blob/a3b81d7bfb3dfbe350a5148dd919ba7f5ce74d59/examples/simulation_pytorch/main.py#L32 

If you want to use more GPUs then you need to change the code inside the `train` function here: https://github.com/adap/flower/blob/a3b81d7bfb3dfbe350a5148dd919ba7f5ce74d59/examples/simulation_pytorch/utils.py#L29

And that will depend on your particular network, etc...

**Comment by Mirian-Hipolito** [comment]
@pedropgusmao thanks for the reply! This is correct, I can only see that the **GPU 0** is being used during training, this is the reason I'd like to take advantage of the rest of GPUs I have available. Is there any sample code for changing the _train_ function accordingly?

Also when I'm running the simulation, apparently is training but all the logs reports the clients results as failures, is this expected? 

![image](https://user-images.githubusercontent.com/94491775/188952733-7d767025-559d-44ec-86e0-1a78e3053822.png)

**Comment by pedropgusmao** [comment]
Hello @Mirian-Hipolito ,

This is just a code example showing the simulation capabilities. Using multiple GPUs is framework-specific and not directly related to Flower. If you want to dive deeper into this topic, you can look into https://pytorch.org/tutorials/beginner/former_torchies/parallelism_tutorial.html .

If you just want to have clients running in different GPUs, maybe you can try changing the previous line to :
`        
self.device = torch.device(f"cuda:{int(self.cid)%4}" if torch.cuda.is_available() else "cpu")
`
Notice the number `4`, this will try and send clients to different gpus depending on their client_id (cid)

**Comment by pedropgusmao** [comment]
Let me know if that works

**Comment by Mirian-Hipolito** [comment]
@pedropgusmao  I did the change suggested above but seems like still assigning to GPU0. 
![image](https://user-images.githubusercontent.com/94491775/188965713-e3470c5e-2391-47ee-9c04-cc171bc8445f.png)

I'm not sure why, also I can see that Ray recognizes the 4 GPUs but seems like it cannot assign them. I've tried 

![image](https://user-images.githubusercontent.com/94491775/188966079-bfc3f63c-7501-4bac-a958-d28d89f178c0.png)

I've tried adding 'num_gpus' key to the client_resources dict but I don't think is recognizing them.

**Comment by pedropgusmao** [comment]
Can you try and add 
`"num_gpus": 1`
to the dictionary in https://github.com/adap/flower/blob/4e7fad99389a5ee511730841b61f279e3359cb16/examples/simulation_pytorch/main.py#L142
and leave:
`self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")` without any index? 
The idea is that Ray will set `CUDA_VISIBLE_DEVICES` will set to a different number (from 0 to 3) for each client and the client will pick the only GPU it sees. 
Let me know if this solves the problem

**Comment by Mirian-Hipolito** [comment]
@pedropgusmao seems like the real issue was that the ray version. I installed flower doing `pip install flwr`, however the version I got for Ray wasn't working with the changes above, I needed to update to 2.0 and this fixed it. Once I implemented your suggestion I can see it's working with multiple GPUs, thanks! The issue that I see now is that the GPU memory is being overloaded  as the iterations are going through .. 

![image](https://user-images.githubusercontent.com/94491775/188997320-8c8eccd4-5c39-49e4-982d-7dfa37546d7c.png)

It seems like this is a well known issue on Ray (https://docs.ray.io/en/latest/ray-core/tasks/using-ray-with-gpus.html#workers-not-releasing-gpu-resources) that can be solved by adding the `max_calls=1`  parameter, but I'm not sure this is the best solution though.

**Comment by Mirian-Hipolito** [comment]
@pedropgusmao sure, thanks for the advice!
